codec-http2/src/main/java/io/netty/handler/codec/http2/Http2FrameCodec.java

```diff
@@ -16,329 +16,506 @@
+        Http2Connection connection = connection();
+        if (connection.isServer()) {
+            tryExpandConnectionFlowControlWindow(connection);
+        }
+    }
+
+    private void tryExpandConnectionFlowControlWindow(Http2Connection connection) throws Http2Exception {
+        if (initialFlowControlWindowSize != null) {
+            // The window size int the settings explicitly excludes the connection window. So we manually manipulate the
+            // connection window to accommodate more concurrent data per connection.
+            Http2Stream connectionStream = connection.connectionStream();
+            Http2LocalFlowController localFlowController = connection.local().flowController();
+            final int delta = initialFlowControlWindowSize - localFlowController.initialWindowSize(connectionStream);
+            // Only increase the connection window, don't decrease it.
+            if (delta > 0) {
+                // Double the delta just so a single stream can't exhaust the connection window.
+                localFlowController.incrementWindowSize(connectionStream, Math.max(delta << 1, delta));
+                flush(ctx);
+            }
+        }
+    }
+
+    void handlerAdded0(@SuppressWarnings("unsed") ChannelHandlerContext ctx) throws Exception {
+        // sub-class can override this for extra steps that needs to be done when the handler is added.
     }
 
     /**
      * Handles the cleartext HTTP upgrade event. If an upgrade occurred, sends a simple response via
      * HTTP/2 on stream 1 (the stream specifically reserved for cleartext HTTP upgrade).
      */
     @Override
-    public void userEventTriggered(ChannelHandlerContext ctx, Object evt) throws Exception {
-        if (!(evt instanceof UpgradeEvent)) {
+    public final void userEventTriggered(ChannelHandlerContext ctx, Object evt) throws Exception {
+        if (evt instanceof Http2ConnectionPrefaceWrittenEvent) {
+            // The user event implies that we are on the client.
+            tryExpandConnectionFlowControlWindow(connection());
+        } else if (evt instanceof UpgradeEvent) {
+            UpgradeEvent upgrade = (UpgradeEvent) evt;
+            try {
+                onUpgradeEvent(ctx, upgrade.retain());
+                Http2Stream stream = connection().stream(HTTP_UPGRADE_STREAM_ID);
+                // TODO: improve handler/stream lifecycle so that stream isn't active before handler added.
+                // The stream was already made active, but ctx may have been null so it wasn't initialized.
+                // https://github.com/netty/netty/issues/4942
+                new ConnectionListener().onStreamActive(stream);
+                upgrade.upgradeRequest().headers().setInt(
+                        HttpConversionUtil.ExtensionHeaderNames.STREAM_ID.text(), HTTP_UPGRADE_STREAM_ID);
+                new InboundHttpToHttp2Adapter(connection(), decoder().frameListener())
```

Comment:
https://github.com/netty/netty/pull/6896#discussion_r132101332 was collapsed ...

> creation of this `ChannelInboundHandlerAdapter` outside the scope of a pipeline or channel lifetime may lead to initialization issues in the future if `InboundHttpToHttp2Adapter` assumes it will live in a pipeline in the future. I wonder if we should not make it a `ChannelInboundHandlerAdapter` for now ... or have it follow the typical life cycle events.

@normanmaurer 
> let me see if I can factor out the stuff a bit...